core/storage/src/trie.rs

```diff
@@ -0,0 +1,926 @@
+        result
+    }
+}
+
+impl<'a> Iterator for TrieIterator<'a> {
+    type Item = TrieItem<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        enum IterStep {
+            Continue,
+            PopTrail,
+            Descend(Result<Box<TrieNode>, String>),
+        }
+        loop {
+            let iter_step = {
+                self.trail.last_mut()?.increment();
+                let b = self.trail.last().expect("Trail finished.");
+                match (b.status.clone(), &b.node) {
+                    (CrumbStatus::Exiting, n) => {
+                        match n {
+                            TrieNode::Extension(ref key, _) => {
+                                let l = self.key_nibbles.len();
+                                self.key_nibbles.truncate(l - key.len());
+                            }
+                            TrieNode::Branch(_, _) => {
+                                self.key_nibbles.pop();
+                            }
+                            _ => {}
+                        }
+                        IterStep::PopTrail
+                    }
+                    (CrumbStatus::At, TrieNode::Branch(_, value)) => {
+                        if value.is_some() {
+                            let value = value.clone().expect("is_some() called");
+                            return Some(Ok((self.key(), DBValue::from_slice(&value))));
+                        } else {
+                            IterStep::Continue
+                        }
+                    }
+                    (CrumbStatus::At, TrieNode::Leaf(_, value)) => {
+                        return Some(Ok((self.key(), DBValue::from_slice(value))))
+                    }
+                    (CrumbStatus::At, TrieNode::Extension(_, child)) => {
+                        let next_node = match child {
+                            NodeHandle::Hash(hash) => self.trie.retrieve_node(hash).map(Box::new),
+                            NodeHandle::InMemory(node) => Ok(node.clone()),
+                        };
+                        IterStep::Descend(next_node)
+                    }
+                    (CrumbStatus::AtChild(i), TrieNode::Branch(children, _))
+                        if children[i].is_some() =>
+                    {
+                        match i {
+                            0 => self.key_nibbles.push(0),
+                            i => {
+                                *self.key_nibbles.last_mut().expect("Pushed child value before") =
+                                    i as u8
+                            }
+                        }
+                        let next_node = match &children[i] {
+                            Some(NodeHandle::Hash(hash)) => {
+                                self.trie.retrieve_node(&hash).map(Box::new)
+                            }
+                            Some(NodeHandle::InMemory(node)) => Ok(node.clone()),
+                            _ => panic!(), // Wrapper with is_some()
```

Comment:
I would put the comment into a panic message to help future yourself searching for the relevant panic :)